New installation instruction and adder for Global Actor

BlurredChoise opened the pull request.

BlurredChoise committed:
Merge pull request #1 from JPETTomography/master

MR to update with JPEt version

BlurredChoise committed:
Changed installation instruction

BlurredChoise committed:
Global and Local Actor User Guide

BlurredChoise committed:
Corrected model name for pPs and prompt gamma

Alvarness commented on JPET_GATE_Use_Examples/README.md:
```diff
@@ -1,107 +1,113 @@
+Before you start do anything be sure you have installed below listed tools:
+ - cmake >= 3.10.2
+ - ccmake >=3.10.2
+ - gcc >= 7.3.0
```
Isn't this requirement a little too tight? ROOT, Geant and Gate all compile with gcc v5

BlurredChoise replied:
No exactly - indeed you can compile with gcc v5 but this version is really old (from 2014-15). In this period of time gcc increased its compatibility with c++14. Moreover, I'm sure we will start using some solutions from c++17 to make code works better. In this case we will omit potential problems with compilation in  the nearest future.

BlurredChoise committed:
Small changes how few methods collect data about track

BlurredChoise committed:
New class which represents hit - for GlobalActor usage

wkrzemien commented on source/digits_hits/src/GateGlobalActorHit.cc:
```diff
@@ -0,0 +1,198 @@
+ extractDataFromStep( step );
+}
+
+GateGlobalActorHit::GateGlobalActorHit( const GateGlobalActorHit& hit ) 
```
if this is just assigning all fields, then why do you need to define it?

BlurredChoise replied:
Yup - it's unnecessary. Now is removed from this class.

wkrzemien commented on source/digits_hits/include/GateGlobalActorHit.hh:
```diff
@@ -0,0 +1,147 @@
+  G4double keV( const G4double& energy ) const;
+
+ private:
+  std::string mVolumeName;
```
I would initialize all the fields you can here.

BlurredChoise replied:
Done

BlurredChoise committed:
Adder for Global Actor

BlurredChoise committed:
Small changes of few Global Actor Hit methods

BlurredChoise committed:
Global Actor is now enable to use Adder ( merging hits )

BlurredChoise committed:
We should only accept hits with parentID == 0 in this adder

BlurredChoise committed:
Added two command for Global Actor: now is enable to set via macro an usage of Adder in Global Actor

BlurredChoise committed:
Added archiving metod for volume name

BlurredChoise committed:
Added copyrights

BlurredChoise committed:
Added class descriptions

BlurredChoise committed:
Removed unused method. Added saving data from adder for each end of event

BlurredChoise commented:
Added new solutions for Global Actor: 
- adder ( merging hits from multiply scattering in the same scintillator)
- hit ( represent data from G4Step which Global Actor uses
- support for end of event ( to prevent data lost for last event in run when we use adder )

Moreover I added copyrights for each file.

BlurredChoise committed:
Added variables initialization

BlurredChoise committed:
Removed copy constructor

wkrzemien commented on source/digits_hits/src/GateGlobalActorAdder.cc:
```diff
@@ -0,0 +1,80 @@
+ hit_to_update.setMomentumDirectionAfterProcess( hit_source.getMomentumDirectionAfterProcess() );
+ hit_to_update.setPolarizationAfterProcess( hit_source.getPolarizationAfterProcess() );
+
+ hit_to_update.setInteractionTime( hit_to_update.getInteractionTime() + hit_source.getInteractionTime() );
```
Just to be sure.
getInteractionTime() returns relative or absolute value?

BlurredChoise replied:
It is always not negative value. GateGlobalActorHit class gets this value from step (G4Step class) by method GetDeltaTime(), which returns difference of local time from post and pre step.
Here is a link to official implementation of this method: 
http://www.apc.univ-paris7.fr/~franco/g4doxy/html/classG4Step.html#1533103052956bdb1e845b0c1054d5c3

wkrzemien commented on source/digits_hits/src/GateGlobalActorAdder.cc:
```diff
@@ -0,0 +1,80 @@
+
+GateGlobalActorAdder::~GateGlobalActorAdder() {}
+
+void GateGlobalActorAdder::processHit( const GateGlobalActorHit& hit )
```
Here some comment how this function work might be useful.

BlurredChoise replied:
Done

wkrzemien commented on source/digits_hits/src/GateGlobalActorAdder.cc:
```diff
@@ -0,0 +1,80 @@
+
+void GateGlobalActorAdder::processHit( const GateGlobalActorHit& hit )
+{
+ if ( hit.getParentID() != 0 ) { return; }
```
what does it mean?
Why parentID !=0?

BlurredChoise replied:
Because  I want to be sure that we only accept data from particles emitted from source. Example: gamma is emitted from source ( parentID=0), then scatter. During scattering energy deposition is transform to electron (parentID !=0).

I added comment to code about this.

wkrzemien replied:
What about the case when the gamma undergoes single Compton in one detector then reacts in the second layer and you have multiple scattering then? 
Shouldn't you also merge that cases?

Also, shouldn't you also merge energies from electrons (I think Daria does it)?

BlurredChoise replied:
It doesn't matter.
We have this scenario:
1. gamma scatter once in scintillator A
2. then scatter multiply in scintillator B
As a results adder will generate two hits: hit 1 in scitillator A and 1 hit in scintillator B ( merged hits ). In both cases we have the the same gamma ( parentID=0 and for example trackID=1).

About electron: 
I don't think so - because it generates a lot of problems:
1. in this case how we handle a post polarization and momentum ?
2. now let's think about this situation:  gamma scatters with 100 keV energy deposition and this is a transferred kinetic energy to an electron. This leads us to situation when total deposited energy is 200 keV right? but this energy deposition by electron is caused by msc (multiple scattering).
I think we should only take this under consideration when we use optical photons mechanism -  but we didn't so it not necessary.

In my opinion, let's GateGlobalActorAdder be a tool which only takes under consideration an energy deposition from gamma. If we want to use more complex mechanism ( with including optical photons ) it is better to create new adder dedicated for this e.g. GateGlobalActorAdderOpticalPhotons.

wkrzemien replied:
Ad 1. You assign it from the first Compton scattering.
Ad 2. I dont understand what do you mean. 
Or you calculate the Edep from the multiple scattering of electron, or you calculate it as a difference between Egamma incoming - Egamma outgoing. If your code does the second thing. Then it is fine and you don't need to take into account the electron scattering.

For sure you dont want to take into account optical photons, but it has nothing to do.
So I assume you are calculating it from the gamma energy difference and it is fine

wkrzemien reviewed (commented):
Nice work

BlurredChoise committed:
Few comments about code
